helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -120,24 +121,24 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
   private Optional<AssignableNode> getNodeWithHighestPoints(AssignableReplica replica,
       List<AssignableNode> assignableNodes, ClusterContext clusterContext,
       Set<String> busyInstances, OptimalAssignment optimalAssignment) {
-    Map<AssignableNode, List<HardConstraint>> hardConstraintFailures = new ConcurrentHashMap<>();
+    Map<AssignableNode, List<String>> hardConstraintFailures = new ConcurrentHashMap<>();
```

Comment:
Why should we change to a string? If we already have the HardConstraint, we can capture any error messages upon failure. Alternatively, we can directly print out the error to get an immediate reason for the failure.

No need to change the entire class and constraint interface for this.

Reply:
Currently, `List<HardConstraint>>` just contains the reference of the hardConstraint singleton object and the error message is just a static string, which is not useful. 

For example: `NodeCapacityConstraint`, We log: ` "Node has insufficient capacity"` which seems not useful enough as we dont get which key is failing for capacity and by how much. We have this as part of debug log but with many consumer now onboarded to WAGED and hard constraint issue happening fairly often so its better that its part of log (error) message itself. Also, we dont want to print it always, there are false positive as well, and only  when there is a total failure.